follow up (#6)
aep/aep.github.io: _posts/2019-02-09-back-to-the-future.md

---
layout: post
title: back to the future
---

This is a follow up to [my original rant about futures](http://aep.github.io/rust-async-without-the-noise/).

> Wow, this kinda blew up and there's been alot more attention than anticipated.

Reviewer: ```suggestion
Wow, this kinda blew up and there's been a lot more attention than anticipated.
```

Which only shows that there is a lot of interest in the state of async in rust.

> Besides the usual useless comments on how i am a terrible person (thank you random stranger on the internet),

Reviewer: I suggest not throwing shade back in this way. I'd instead phrasing the same message in a positive way. For example:

> Many people focused on the tone of my original article, rather than the technical claims I made. I received a lot of good feedback, including on some of the proof or technical details mentioned in my original article, which I am very thankful for. The Rust community is amazing.

Generally, understanding intent around tone and emotion in text form sucks, and will generally always be interpreted in the way that makes you look worst.

there was actually alot of very good feedback. Some people went through the entire article and picked it apart
for technical details. The rust community is amazing.

> The biggest point i need to address is about futures 0.3 and how the post doesn't actually offer any learnings

Reviewer: ```suggestion
The biggest point I need to address is about futures 0.3 and how the post doesn't actually offer any learnings
```

Standalone `I` is always capitalized in English.

> for the _new_ futures, since all the reasons i offered why futures are broken only applies to future 0.1.

Reviewer: ```suggestion
for the _new_ futures, since all the reasons I offered why futures are broken only applies to futures 0.1.
```

But before we get to that, i want to reply to a few other comments.

> but it's not at all clear from this blog post how Osaka handles multiplexing connection-oriented sockets.

For completeness, here's how you poll multiple things:

```rust
#[osaka]
fn foo(poll: Poll, a: Stream, b: Stream) {
  loop {
    let again = match a.read() {
      Ready(something) => return something,
      Again(again) => again,
    };
    match b.read() {
       Ready(something) => return something,
       Again(again2) => yield again.merge(again2)
    };
  }
}
```

this is very explicit. It could be hidden behind a macro, but these are easy to build once the basic concept is clear.

> osaka uses more std than futures

> osaka is... not good code. And its unfortunate that this rather rough implementation got mixed 

Reviewer: Consider making this argument at the top of the article, as it is essentially your main point. Something like:

> In general, I provide technical details of Osaka only to illustrate my main point: That there are perhaps better abstractions that could be chosen for Futures and Tokio that would allow for a more flexible implementation across all use cases, including for constrained embedded systems.

into the actual point i'm trying to bring to the rust core devs that futures is the wrong interface.

I cannot implement osaka the way that it should be implemented without forking rustc. 

> Hence i was hoping for  the attention of rust devs to implement async as a generic interface rather than as future,

Reviewer: ```suggestion
Hence I was hoping for the attention of rust devs to implement async as a generic interface rather than as a future,
```

so it becomes reusable.

> A good interface would be Fn->Fn.  In a perfect world like the clay programming language, 

Reviewer: Consider linking to the Clay programming language

generating yield points for Fn->Fn is part of the core language like so:

```
async string something(Socket *socket) {
    Packet packet = with socket.receive();
    return packet.decode();
}
```

> In rust that is a Generator. Almost. Since generators can't take arguments, 

Reviewer: ```suggestion
In Rust that is a Generator. Almost. Since generators can't take arguments, 
```

we need to wrap the continuation arguments into much more complex structures, such as channels.
It's more complex and we need more trickery to implement the same thing, 

> which futures.rs gracefully pushes onto tokio in a symbiosis that led to std::futures effecively implementing an interface for tokio.

Reviewer: When you make a claim that is a negative, provide an example of what Tokio does, so you can demonstrate how Osaka solves the same issue in a better way. This allows you to SHOW your claim, instead of just making it, or having to state "Look, Osaka does this better".

osaka unlike futures actually does this

```rust
impl Future for X
where X: std::ops::Generator
```

> so you *can* use any generator as a future, and its something futures.rs should be doing rather than pushing futures into std with a type signature that dictates how you need to implement the execution engine (hint: it needs to be like tokio)

Reviewer: ```suggestion
so you *can* use any generator as a future, and it's something futures.rs should be doing rather than pushing futures into std with a type signature that dictates how you need to implement the execution engine (hint: it needs to be like tokio)
```

Be careful about using words like "should", as it is then up to you to prove this claim. Consider weakening the claim a bit, like "and it is something I believe futures.rs *should* be doing, rather than...".

> There are example of embedded execution engines that are not tokio

> No there are not. All the ones that have been linked in comments are re-implementations of tokio. 

Reviewer: This paragraph takes a very combative tone that is unlikely to win you any favors in your arguments. Consider instead something like:

> While there are proof of concept implementations including X and Y, these are still in a proof of concept stage, and many are a 1:1 mapping of what Tokio does. In contrast, Osaka is a library that I have used extensively in production, across x0,000 resource constrained embedded devices, and N separate libraries and applications.

This avoids:

1. insulting the other implementations
2. insulting the people who have developed these implementations
3. insulting the people who have suggested you look at these implementations.

I find it also very frustrating that litteraly all of them are "toy implementations" 
and the argument is that "futures was designed for this use case", but zero of these authors have any background in embedded.

> I can surely apprechiate that the rust core team in good faith is trying to make embedded work, 

Reviewer: Again, this paragraph is combative (and has a couple spelling errors).

I suggest something like:

> I appreciate that the design decisions that have been made were intended in good faith to be sound for resource constrained targets, however I have been somewhat frustrated that my arguments, based on concrete experience have been countered with the claim of "it works in theory", without any data to back it up.

but consistently dismissing feedback from people who actually ship real products because "it works in theory" is pretty infuriating.

> Yet for the sake of objectively arguing with strangers on the internet i'm actually going to implement an execution engine for futures 0.3 in this post.

Reviewer: Again, combative.

> For the reasons I have listed above, I would like to demonstrate the issues that I have raised, with concrete data driven examples, to illustrate exactly the issues that I have seen in practice.

futures under realistic constraints
===============================

> While i can rage about tweets that dismiss embedded as an uninteresting fringe from the rust core team all day, 

Reviewer: Combative.

> I'd like to pick a couple metrics that are critically important to my use cases, and demonstrate these comparisons objectively.

there's actually a rather large group of readers out there who care about objective facts. 
So i'm picking some real life constraints here to benchmark against:

**flash size**

> We'll pick two targets: a mips24k board running linux with 4Mb flash (hint: that's very small) and a cortex m0 with 256kb of ram (hint: that is pretty big). 

Reviewer: Sarcasm doesn't work well over written form/the internet.

**runtime/os**

on the mips target we do not have GNU libc for license reasons (and because its massive). we'll roll with the vendor default frankenstein libc.

on the cortex m we have no threading primitives at all because we need to run cooperatively with the vendor binary code which does radio telemetry.

a slow start
============

Starting with arm i created a simple fake async resource:

```rust
struct Later(u32);
impl Future for Later {
    type Output = u32;
    fn poll(self: std::pin::Pin<&mut Self>, wk: &std::task::LocalWaker) -> Poll<u32> {
        Poll::Ready(self.0)
    }
}

async fn boop(beep: Later) -> u32 {
    let a = await!(beep);
    a * 2
}

fn main() {
    let then = boop(Later(32));
}

```

> that's a reasonable interface until here/ Pretty close to Fn->Fn

Reviewer: This sentence doesn't really read well. I'm not sure what you're trying to convey here.

> **Except all of this is inside std, so no embedded!**

Reviewer 2: Huh? You omitted the `use std::future::{Future};` part from the example, but so far it [actually works](https://play.rust-lang.org/?version=nightly&edition=2018&gist=9b504f22a634af896ab14d7ed0e2f2a2) with `core` instead of `std` (`future`, `pin`, `task` are all part of `core`).

As a generic comment: using `Future` on embedded imho requires static storage for tasks (they must not move and Box isn't available) and wakers (you need to implement ref counts for them, but there is no heap), but I don't consider that a major problem.

Author: huh, i completely missed that. gotta redo the whole post then. thanks!

```
error[E0433]: failed to resolve: use of undeclared type or module `std`
 --> src/main.rs:4:5
  |

> 4 | use std::future::{Future};

Reviewer: I think this is exposed as `core::future::Future`. See [here](https://doc.rust-lang.org/nightly/core/future/trait.Future.html).

Reviewer: This is common in bare metal embedded development. Many items like this are actually defined in core, but re-exported in std. See also things like `core::result::Result` and `std::result::Result`.

  |     ^^^ use of undeclared type or module `std`

```

I wasn't expecting to give up this early, considered the repeated claim that futures is designed for embedded,
but maybe i'm doing it wrong?

moving on to mips/linux then, which at least has an std implementation.

> the next thing we need is sort of poll that thing. Could be pretty easy, since all we need is this waker thing to call poll.

Reviewer 2: The `UnsafeWake` trait is unsafe, and most people have a hard time correctly using unsafe, so if you have std available `local_waker` is the right way to go, yes. Embedded is hard, so don't be surprised - but it does support embedded if you need it.

If you have a single-threaded environment it should be easier to implement them correctly :)

> Apparantly it you're supposed to construct it from this function:

Reviewer: This sentence doesn't make sense grammatically. Probably:

```suggestion
Apparently you're supposed to construct it from this function:
```

```
pub unsafe fn local_waker<W>(wake: Arc<W>) -> LocalWaker

```

> Arc is a pretty scary requirement, since that requires locking primitives to workd.

Reviewer: Arc doesn't necessarily require locks. See also `core::sync::atomic::Atomic*`.

```suggestion
Arc is a pretty scary requirement, since that requires locking primitives to work.
```

An Arc is also not 'zero cost', depending on your cpu (bad on mips) but moving on, trying to implement a "Wake" since that's required. 

But we're getting stuck here:

```
38 |     Pin::new(&mut then).poll(&wk);
   |     ^^^^^^^^ within `impl std::future::Future`, the trait `std::marker::Unpin` is not implemented for `std::future::GenFuture<[static generator@src/main.rs:17:35: 20:2 beep:Later {Later, ()}]>`
```

> Well, I guess you can't pin stack values? No idea why, but apparantly it works fine when you move it on the heap:

Reviewer 2: Yes, boxing (and then pinning) a value is an easy way to make sure it (the inner part) doesn't get moved anymore. Stack values can usually be moved around, and you will need to use unsafe methods to create a `Pin` for those (and you need to ensure they are not moved afterwards: not sure if that is even possible for stack values - the compiler might be free to move them anyway. A static should do fine though).

You could try to write a generator that creates states that are `Unpin`, then you can freely move the state between invocations; but you can't keep references to you own state between calls... The combinators you so much dislike are btw all `Unpin`...

While I understand that `Pin` is not an easy concept to grasp (having troubles myself with it), I don't see an alternative and think it is a pretty nice idea.

Author: pin confuses me a lot. 
why don't generators not just keep their state local on the stack?

Reviewer 2: Which stack? The generator can't use the normal stack, and a separate stack would need to be allocated first - but we don't want to rely on having an allocator.

So: the "generator literal" (`|| { ... yield ... }`) creates a "state" object, and implements the generator trait for it.  This state simulates the stack of the closure (but only the part that needs to be kept across resumes; it has a fixed size! Calling normal methods from the generator uses the standard stack).

Now assume you have `let mut buf = [0u8; 1024]; await!(sock.read(&mut buf))` in your async code; the `sock.read` call will create a Future, but its lifetime is limited to the `&mut buf` lifetime, which is bound to the generator state itself.  You must not move the buffer, and therefore the generator state, while the read is pending.  This is what `Pin` solves.

```rust
    let mut then = Box::pin(boop(Later(32)));
    let wk = unsafe { std::task::local_waker(std::sync::Arc::new(Wakething{}))};
    println!("{:?}", Pin::new(&mut then).poll(&wk));
```

> not sure how that falls in line with "better than callbacks because fewer pointer indirections", but i'm not sure if that matters either.

Reviewer 2: Not sure where this quote is from, but obviously "waking" a waker is always a dynamic dispatch to a "callback", which should only add the task to the execution queue and not do anything important, but it's still a callback.

Author: dont remember 100% certain so i dont want to name the one who said it.
but it was a big argument to why tokio is supposed to be better than Kite.

Reviewer 2: I don't know about Kite; my guess would be that nesting async operations uses dynamic dispatch on every level; with futures you have one (dynamic) callback for the most inner "blocking" operation in your task. Also there shouldn't be many codepaths for those callbacks, so they might be in the cache most of the time (their associated state probably won't be).

interrupting all the focuses
============

> next thing we want to do is implement a microcontroller wakeup scenario. We can't actually test it because fututures doesn't actually compile for arm.. but we'll at least validate the claim that it was designed for embedded in theory.

Reviewer: Again, this claim is untrue. See `core::future::Futures`.

> Assuming you somehow managed to implement all the thread primitives for your platform, which is tons of work, we'll now need to wakeup tasks from real interrupts somehow.

Reviewer: Also, this is another "toy implementation", but you might want to check out [embrio-rs](https://github.com/Nemo157/embrio-rs), which at least is a reasonable toy implementation of Tokio for bare metal systems.

You will likely have people pointing this out to you (it comes from the Networking-WG), so you should be prepared to address this.

we'll simulate an interrupt like this:
```rust
let external_event = thread::spawn(||{
    thread::sleep(time::Duration::from_millis(10));
    interrupt();
});
```

Getting a future to complete when the interrupt triggers, is easy in clay:

```
global future = some_future_stack;
interrupt(event) {
    future(event)
}
```

> since rust does not allow us to pass arguments into generators (or futures), we need to use a channel.

Reviewer 2: I'm sure that's not what you meant, but obviously async functions can take parameters (generators don't strictly speaking, but you can always move data into the closure). Maybe you want to clarify what actually is your point here, because "we can't build (single shot) channels without building (single shot) channels" is not an interesting argument.  (Implementing a single-threaded single shot channel isn't that hard; multi-threaded and handling back-pressure makes it interesting.)

Author: interesting that its not obvious. I should expand this.  How about an example:

if generators allowed continuation arguments you could just do

let a = || {
    let b = yield;
};

a(42);

Reviewer 2: While being able to pass data to the resumption of a generator would make it easier to implement async/await without TLS I don't think it'd solve your problem: the generator is owned by the executor, which requires the same generator type across all tasks - so all tasks would need to use the same continuation arguments.

I still think it would be nice to support continuation arguments, and I'd probably prefer the normal argument syntax `|b| { yield }`: all resumes (including the first one) need to pass the same arguments, and there lifetime should by default be limited to the "resume cycle".

std::sync::mpsc::channel is nice, and we can easily use try_recv inside pollErr

```rust
struct Interrupt (mpsc::Receiver<u32>);
impl Future for Interrupt {
    type Output = u32;
    fn poll(self: std::pin::Pin<&mut Self>, wk: &std::task::LocalWaker) -> Poll<u32> {
        match self.0.try_recv() {
            Ok(t)  => Poll::Ready(t),
            Err(_) => Poll::Pending,
        }
    }
}

```

if we poll this thing, it returns immediately with "Pending" without any additional context what it is waiting for.
This is a major difference to osaka.
We need some way to know when to call poll() again. Ideally, we'd just call it from interrupt() because we know its ready then,
or we could return some sort of token in Pending to tell an external caller that we're waiting for the channel to become ready.

Waker does not let us access the inner type either, so

> **the only available option is to do exactly what tokio does**

Reviewer 2: I'm pretty sure you don't need thread_local to implement an executor.  You just have to store LocalWaker (or Waker) somewhere your interrupt code can find it.

Author: i dont see how else you can do this. waker doesnt let you register anything to an eventloop

```rust
thread_local! {
    static REACTOR: Arc<Reactor> = Arc::new(Reactor::new());
}

fn poll(self: std::pin::Pin<&mut Self>, wk: &std::task::LocalWaker) -> Poll<u32> {
    REACTOR.with(|reactor| reactor.register(&self));

```

> note the implementation of await!:

Reviewer 2: I'm pretty sure this could be implemented without TLS in the long run.  Also if you're only single-threaded TLS should be easy :P

Author: sure, i could port my linker to the vendor libc and emit fake TLS.

but i could also just use osaka, which doesnt need TLS :P

Reviewer 2: I just managed to build an alternative `async`/`await` mechanism which is compatible to futures and doesn't use TLS, see https://github.com/stbuehler/rust-cofutures-poc

```rust
macro_rules! await {
[..]
$crate::future::poll_with_tls_waker(unsafe {
    $crate::pin::Pin::new_unchecked(&mut pinned)
})
[..]
```

> requires TLS anyway. So if you do not have a working TLS implementation, or don't want to afford the extreme overhead, you're pretty screwed anyway.

Reviewer: This sentence has a combative tone.

In my case rustc emits incorrect TLS frames for static binaries on mips24k, so it crashes immediately.

> On microcontrollers, TLS is very hard to stabilize. [I implemented a linker](https://github.com/aep/elfkit) and can tell tales of how fricking broken TLS is. The whole idea is insane.

Reviewer: Extremely combative tone.

> We'll move on simulating everything on x86_64 , which is clearly the only platform that futures is intended for.

Reviewer: ```suggestion
We'll move on simulating everything on x86_64, which seems to be the platform that futures is most heavily developed against.
```

standing at the edge
====================

Still, the claim that you CAN implement a different behaviour than tokio stands on being able to implement level triggered behaviour,

> which is simpler, more stable and easier to work with, rather than the edge triggered that tokio enforces (because its easier to port to windows apparantly).

Reviewer: Claims without proof. I am not personally aware of the differences between edge and level triggered behavior (at least in this context), so I have no idea whether your claims are true or false.

We'll try with a simple linux udp server using mio.

```rust

lazy_static! {
    static ref POLL : std::sync::Arc<mio::Poll> =  std::sync::Arc::new(mio::Poll::new().unwrap());
}

struct Udp{
    sock: UdpSocket,
}

impl Future for Udp {
    type Output = Vec<u8>;
    fn poll(self: std::pin::Pin<&mut Self>, wk: &std::task::LocalWaker) -> Poll<Vec<u8>> {

        POLL.register(&self.sock, mio::Token(0), mio::Ready::readable(), mio::PollOpt::level()).unwrap();

        let mut b = vec![0;1024];
        match self.sock.recv_from(&mut b) {
            Ok((len, _))  => {
                b.truncate(len);
                Poll::Ready(b)
            }
            Err(_) => Poll::Pending,
        }
    }
}

```

first attempt: not so good. you can of course not re-register to poll.

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 17, kind: AlreadyExists, message: "File exists" }', src/libcore/result.rs:997:5
```

> tokio solves this with an extreme amount of overhead per wakeup. We'll try to be a little better by moving the registration into the constructor.

Reviewer: Prove that Tokio has extreme amounts of overhead per wakeup. You do not demonstrate this claim anywhere.

```rust
impl Udp{
    pub fn new() -> Self {
        let sock = UdpSocket::bind(&"127.0.0.1:11111".parse().unwrap()).unwrap();
        POLL.register(&sock, mio::Token(0), mio::Ready::readable(), mio::PollOpt::level()).unwrap();
        Self{sock}
    }
}

```

and voila, it works.

```
Pending
Ready("bla\n")
```

trying to do it in a loop is a little confusing, since this doesnt work:

```rust
async fn boop(beep: Udp) {
    loop {
        let a = await!(beep);
                       ^^^^ value moved here, in previous iteration of loop
```

> apparantly you need to clone it, or use an iterator interface, so more allocations for the 'zero cost much better than fn->fn' thing,

Reviewer 2: It should already work with `await!(&mut beep)` (although usually Futures shouldn't be polled after they were ready once; so you should create a new Future for reach receive operation, e.g. a `struct UdpReceive<'a>(&'a mut Udp)`).

Author: doesnt compile.

creating a new future sounds like alot of overhead.

Reviewer 2: Yes, `beep` needs to be mutable too, which the compiler should have told you (even how to do it). A little bit more effort from your side would be appreciated. See my full example for working code.

A new future is not "a lot of overhead", it's the correct abstraction: your Udp socket is not only made for receiving data, it should also be able to send data, so you will have to build something for that. And a newtype doesn't cost memory, and the compiler should inline most of the trivial code, so don't complain about performance overhead unless you actually measured it.

> apparantly you need to clone it, or use an iterator interface, so more allocations for the 'zero cost much better than fn->fn' thing,

Reviewer: Combative tone.

but finally i got it working. here is the full code: [https://gist.github.com/aep/bd44329ddc58c81efb242932fa59aece](https://gist.github.com/aep/bd44329ddc58c81efb242932fa59aece)

> So it is possible.

Reviewer 2: And a lot of it is possible without `std` (although `await!` and `async fn ...` isn't there yet): https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ae20bc0d896ece19ee076e84b0398d0a

Author: hey thanks for that! will continue with your example

conclusion
===========

unfortunately there was no way to benchmark futures on the target devices for real, since it doesnt even compile. We'll compare the theoretical impact

> |                      | tokio        | futures 0.3 | osaka      | fn->fn      |

Reviewer 2: I'm not quite sure what you think `fn -> fn` is; if it is a "callback" which will return a new callback as continuation, it will certainly involve state, and will have to copy all state from the old to the new callback (at least the bits it still needs).
Can't see how that would be possible in a sane way without a heap (or allocating and using a separate stack, which is basically the same as allocating on the heap).

Author: yeah, separate stack.
i'm not sure how to implement it in rust, but in clay we just used a stack of lambdas.

i.e.
```
|| {
    let b = yield;
    let c = yield;
}
```

is equivalent to

```
|| {
    |b|{  
         |c| {
         }
    }
}
```

and yes that state is obviously on a separate stack, but that's not equivalent to heap because you can limit stack at compile time.

Reviewer 2: And what happens if I want two "udp sockets" (or whatever)? How can the compiler know how many parallel instances I'll need, and when to pick which? This will come with lots of restrictions (which I would refuse to work with).

I admit you won't need to copy data between the lambdas, as the inner one simply can access the outer data by reference.

> |                      | tokio        | futures 0.3 | osaka      | fn->fn      |

Reviewer: This table provides very little actual evidence for your claim. I highly suggest you create a repo containing the code project examples used to determine the items in this table, with numeric data (`.text` size, etc.), rather than "is good" or "sucks" or "extreme overhead". Otherwise it is hard to claim you are making a data driven experiment, and your points are all subjective.

Otherwise you are likely to win anyone else over to your point, which is sort of the whole point of your follow up post.

|----------------------|--------------|-------------|------------|-------------|
| size bloat           | refference   | smaller     | smaller    | zero        |
| activation overhead  | extreme      | extreme     | still bad  | single jump |
| works on cortex m    | no           | no          | not yet    | yes         |
| works on mips        | no           | no          | yes        | yes         |

> | TLS singleton        | yes          | yes         | no         | no          |

Reviewer 2: Only the async/await code generation in `futures 0.3` requires TLS for accessing the waker.

Author: but thats my whole point. :D

if async just emitted a generic type, all of the other discussion was a non issue.

Reviewer 2: Then https://github.com/stbuehler/rust-cofutures-poc solves all of your problems?

> | per activation alloc | yes          | yes         | no         | no          |

Reviewer 2: Neither activation (calling wake) nor "cloning" a waker requires allocation in "futures 0.3"; the `std` recommended `Arc`-way of constructing a waker requires allocation, but you only need to allocate it once per task.

Author: you need to clone the thing you poll, because of its weird api.

Reviewer 2: No you don't. Do you see any `clone` in my example?

> | push model           | no           | no          | yes        | yes         |

Reviewer 2: Maybe you could explain what "push model" is for you?

Author: i was hoping i did further up. thanks for letting me know i wasnt clear.

but now that i can compile no-std, i might as well add a real example
